docs/pages/pmd/userdocs/cli_reference.md

```diff
@@ -87,26 +97,35 @@ The tool comes with a rather extensive help text, simply running with `--help`!
                             the given language `&lt;lang&gt;`. Parsing errors are ignored and unparsable files
                             are skipped.
                             
+                            <p>`--use-version` is still needed to specify the version of the language to use</p>
+
                             <p>This option allows to use the xml language for files, that don't
                             use xml as extension. See [example](#analyze-other-xml-formats) below.</p>"
     %}
     {% include custom/cli_option_row.html options="--ignore-list"
                option_arg="filepath"
                description="Path to file containing a list of files to ignore, one path per line.
-                            This option can be combined with `--dir` and `--file-list`.
-                            This ignore list takes precedence over any files in the file-list."
+                            This option overrides files included by any of `--dir`, `--file-list` and `--uri`."
     %}
-    {% include custom/cli_option_row.html options="--help,-h,-H"
+    {% include custom/cli_option_row.html options="--help,-h"
                description="Display help on usage."
     %}
-    {% include custom/cli_option_row.html options="-language,-l"
-               option_arg="lang"
-               description="Specify the language PMD should use. Used together with `-version`. See also [Supported Languages](#supported-languages)."
+    {% include custom/cli_option_row.html options="--use-version"
+               option_arg="lang-version"
+               description="The language version PMD should use when parsing source code.
+                            <p>Values are in the format of *language[-version]*. If version is omitted,
+                            or *-latest* is used the latest supported version is used.</p>
```

Comment:
I don't really like this specification... This is supposed to select a language version. `java` is not a language version it's a language. Specifying `--use-version java` looks bad and also is a noop, as that is the same as not specifying `--use-version` at all (which uses the default version).

Also I don't think the `-latest` suffix has a use case. For nearly all languages the default version is the only version, so `--use-version` is not useful anyway. For those languages which have several versions, the default is usually the latest, and when that's not the case (java), then it's for good reason: we want people to specify preview versions explicitly as they are unstable.

This use of the language name as a version and of the `-latest` suffix also introduces inconsistencies with how language versions are parsed in the ruleset xml (there we only support explicit language versions).

Please remove support for `-latest` in this PR. If you want this feature to be reconsidered later then please open an enhancement request, but let's keep this PR minimal.

Reply:
Thanks for voicing your concerns. I've some concerns on this flag myself, although on a different area, so I'm happy to discuss it here.

> For nearly all languages the default version is the only version, so `--use-version` is not useful anyway.

That's not completely true… Even for languages where we support a single version (`apex-54`, `ecmascript-ES6`) having it here serves several purposes:
- Firstly, this is the only place where we clearly state the version being used for single-version languages beyond release notes.
- Being able to explicitly state the intended version works as a failsafe, allowing for an error to be detected if you upgrade PMD and the supported version is changed in a non-backwards compatible way (ie: we had 4 incompatible bumps of the Swift grammar… breaking changes are not even limited to major versions, Swift 2.0 and 2.3 had major incompatible grammar differences)
- It provides clarity. Looking at `--use-version` help / autocomplete, and not finding a language listed at all is misleading as to wether or not PMD actually supports that language. Having it there removes any doubt.

Therefore, I strongly believe that a user should be able to explicitly reference any supported language (just like it was done when it was 2 separate flags). I hope you are able to agree on this.

Having said that, let me walk you through my reasoning to getting to the current approach:

Unfortunately, we have languages where we do have versions (and aliases) even if only 1 (`java`, `ecmascript`, `apex`, etc.) and others where the version is not set at all (`vm`, `xslt`, `xml`, `html`, etc.). We either force languages to set an explicit version (`xml-1.0`, `html-5`, `xslt-3.0`), or we end up with "short" versionless names being accepted (and using the default version).

I personally refrained from the first one because there are several languages where the implementation is not necessarily appropriate for a given version. For instance, our "html" grammar can look into markup written HTML 4.01, HTML 5, XHTML 1.1… but is lenient, accepting things that would not be valid in those versions. Even worst is Velocity (vm), Apache Velocity versions the engine, not the grammar itself… what version do we state to support?

So, if we are to have "versionless" languages, why would `vm` be valid and use the default version, but `java` not do the same? I added that in favor of consistency.

- do we agree that being able to specify a language version **explicitly** is desired even if that's the only version supported?
- if we do, is there another way around the "versionless" languages I mentioned?

Reply:
Hey Juan, thanks for your response.

> * do we agree that being able to specify a language version explicitly is desired even if that's the only version supported?

I agree that this can be useful to have a failsafe, as you explain in point 2. However I'm not sure how the technical solution you propose helps (having `swift` be an alias for the default version of the language).

Assuming `--use-version swift` selects the default version, then it will not blow up when the Swift parser is changed incompatibly. It will also not blow up when we change the default version, which we routinely do in minor releases. Again, this is a perfect noop and having that in the autocomplete, for every language, can only be confusing IMO.

> We either force languages to set an explicit version (xml-1.0, html-5, xslt-3.0), or we end up with "short" versionless names being accepted (and using the default version).

> I personally refrained from the first one because there are several languages where the implementation is not necessarily appropriate for a given version. For instance, our "html" grammar can look into markup written HTML 4.01, HTML 5, XHTML 1.1… but is lenient, accepting things that would not be valid in those versions. Even worst is Velocity (vm), Apache Velocity versions the engine, not the grammar itself… what version do we state to support?

I think the first solution is the only one that provides a failsafe, and the information you suggest be documented (your points 1 and 3). That's the only way we can *explicitly* set a language version (`--use-version swift` is decidedly not explicit).

I personally think we should absolutely be lenient during parsing. People might be using PMD in an interactive session (eg within an IDE) with half-written code, and an overly strict parser is annoying. Labeling our HTML parser "HTML 5" makes a commitment to being able to parse *at least* HTML 5, but I don't see a problem with it being able to parse other things as long as rules are coded defensively.

The only functional uses of the language version outside of the parser is to filter out rules based on min-max supported versions, and also to have rules have different behavior based on versions of the analysed source. I think therefore, that we should try to detect and report language features that are from another language version, *if* it could help the user select a more specific language version, which could give them access to more specific rules for the version they use.

(In fact, maybe we should have an HTML 4 language version which uses our HTML 5 parser, since you mention it would probably parse correctly. That could be useful to write rules for either version.)

----

Trying to wrap up, 
- I think your concerns, which are all very relevant, are not well served by being able to set the default version on the command line.
- a solution could be to require that versions have an explicit name, like you suggested. I dont see how we can otherwise mention the version name explicitly on the CLI...
- this could however complicate how we handle parser updates during minor releases of PMD. Would we remove the old version? Should we keep the old parser around? Previously we just mentioned it in the release notes
- maybe, versions that are hard to name (eg VTL) could be given a name like `vtl-default`? Version IDs do not have to be a number IMO

(Note that my initial intuition about this all was different, as summarised in #4048. We should update that later with the outcome of this discussion)

Reply:
Thanks!

> * I think your concerns, which are all very relevant, are not well served by being able to set the default version on the command line.
> * a solution could be to require that versions have an explicit name, like you suggested. I dont see how we can otherwise mention the version name explicitly on the CLI...

Don't get me wrong, I 100% believe that we need to add versions for languages that very clearly match that (ie: swift, xml, pom, etc.). I simply stated that not all cases would be so straight forward.

> * this could however complicate how we handle parser updates during minor releases of PMD. Would we remove the old version? Should we keep the old parser around? Previously we just mentioned it in the release notes

This is already an issue, we were simply sweeping it under the rug so far. Keeping the old parser around is not easy because of the tight relationship between grammar > parser > node classes > rule visitors. A Java Rule is tightly bound to a set of node classes, so they would be built for one or the other (or we need to have both sets under the same package / class names, ensuring binary compatibility, and somehow load one set or the other dynamically at runtime). XML rules are easier, as they only require class names and attributes to remain the same, they map names to classes in a given package at runtime.

The only language were we properly support multiple versions is Java, and this entitles 2 things:
- Java's promise on future-compatibility. Any code valid in Java 1 is still valid when compiled under Java 19.
- Manually crafting the parser to include the version checks.

The version checks are probably a nice to have. It only makes parsing stricter; but the first one is critical, and non-dependant on us. Most languages work that way, but not all (again, Swift versions 1 through 3 were constant makeovers, it's been more stable since then).

For those that work, we can add a new version when we update the grammar to be explicit.
For the others, we need to either break compatibility, hold for the next major, or go through a heavier refactor. This is something we probably can skip for now (as we have so far), in the hope that at least for currently supported languages, no further breaking changes are to occur. If it does happen, we may assess our options at that time and make a call.

> * maybe, versions that are hard to name (eg VTL) could be given a name like vtl-default? Version IDs do not have to be a number IMO

Funny you should say that! `-default` was actually my first approach at this. But again, all languages have default versions, why would `vm-default` be valid but not `java-default`? I then swapped to `-latest` just out of trying to convey more meaning into the version, as the default is the latest standard version (no previews, which are not even available without further flags when executing Java / PMD).

Both `--use-version vm-default` and `--use-version java-default` would technically be a noop, simply explicitly stating we want to use the default version.

Reply:
> Don't get me wrong, I 100% believe that we need to add versions for languages that very clearly match that (ie: swift, xml, pom, etc.). I simply stated that not all cases would be so straight forward.

Ok I think I had misunderstood, thanks for clarifying.

Just to go back to the point of my original comment, neither a `-latest` or a `-default` suffix, nor bare language names, help with the stated goals of documentation and providing a failsafe for version upgrades, so I still think we should remove them from this PR. 

> The version checks are probably a nice to have. It only makes parsing stricter; but the first one is critical, and non-dependant on us. Most languages work that way, but not all (again, Swift versions 1 through 3 were constant makeovers, it's been more stable since then).

They are a nice to have, but in the interest of user-friendliness we could make them a warning instead of a parsing error. Detecting features that belong to a different version is done to recommend a more specific version, and we can give this recommendation to the user without crashing the analysis.

> For the others, we need to either break compatibility, hold for the next major, or go through a heavier refactor. This is something we probably can skip for now (as we have so far), in the hope that at least for currently supported languages, no further breaking changes are to occur. If it does happen, we may assess our options at that time and make a call.

Let's do that and skip the topic until it's relevant. Maybe that would be a good topic for an ADR (#4072)

> Funny you should say that! `-default` was actually my first approach at this. But again, all languages have default versions, why would `vm-default` be valid but not `java-default`? I then swapped to `-latest` just out of trying to convey more meaning into the version, as the default is the latest standard version (no previews, which are not even available without further flags when executing Java / PMD).
> Both `--use-version vm-default` and `--use-version java-default` would technically be a noop, simply explicitly stating we want to use the default version.

I was only meaning this for this specific language. I think adding this way to write a noop for all languages would only clutter the help and autocomplete, and add confusion as to what `--use-version` does. Explicitly requesting the default defeats the point of having a default IMO... And again, it is inconsistent with how we write versions in the ruleset XML.

Another proposal for this VTL version, maybe `vtl-0` would work. We would use 0 as the version number because we don't have any meaningful number to use. By not naming it `default` we avoid the users thinking they're setting the default when they're actually setting an explicit version (and extrapolating that there should be a `-default` for all languages). Although I'm not sure we should invest time thinking about this is as the VTL module might be on the way out (#3121). Maybe we should take a decision on that first...

Reply:
ok, let's do that. I'll rework this to only include explicit version names and aliases. You can probably make the version name required as part of #3782. We will update #4048 to reflect this, and we can discuss there which versions should be listed for each currently "versionless" modules.

Are we aligned?

Reply:
Sorry I forgot to respond. Yes this sounds like a good plan to me. I will create a new issue to supersede #4048, so that we don't edit it beyond recognition. The version name can be made required in the PR #4060